[Site Isolation] Touch event regions are missing for elements which have touch handlers but no JS listeners

lilyspiniolas opened the pull request:
#### dd0cc81543c22772b348b5d827c8f256437f1ca7
<pre>
[Site Isolation] Touch event regions are missing for elements which have touch handlers but no JS listeners
<a href="https://bugs.webkit.org/show_bug.cgi?id=302689">https://bugs.webkit.org/show_bug.cgi?id=302689</a>
<a href="https://rdar.apple.com/164939029">rdar://164939029</a>

Reviewed by Abrar Rahman Protyasha.

Switch controls, slider thumb elements, and elements with unaccelerated overflow scrolling all
require synchronous touch event handling but do not have JS listeners. The lack of JS listeners
results in us failing to generate touch event regions for these elements.

Fix this issue by introducing a new EventTargetFlag `HasInternalTouchEventHandling` which
is set for these elements. In addition to targeting elements based on JS listeners in
`Adjuster::computeEventListenerRegionTypes`, we now look for the new EventTargetFlag as well.
If present, we add synchronous event listener region types NonPassiveTouchStart,
NonPassiveTouchMove, NonPassiveTouchEnd, &amp; NonPassiveTouchCancel.

In `SliderThumbElement` we no longer check if a renderer is present before registering for
touch events. Elements without a renderer will not get an event region in the first place.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/EventTarget.h:
* Source/WebCore/html/CheckboxInputType.cpp:
(WebCore::CheckboxInputType::disabledStateChanged):
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::updateTouchEventHandler): Deleted.
* Source/WebCore/html/HTMLInputElement.h:
* Source/WebCore/html/HTMLInputElementInlines.h: Added.
(WebCore::HTMLInputElement::updateTouchEventHandler):
* Source/WebCore/html/InputType.cpp:
(WebCore::InputType::hasTouchEventHandler const):
* Source/WebCore/html/RangeInputType.cpp:
(WebCore::RangeInputType::createShadowSubtree):
* Source/WebCore/html/shadow/SliderThumbElement.cpp:
(WebCore::SliderThumbElement::willDetachRenderers):
(WebCore::SliderThumbElement::didAttachRenderers):
(WebCore::SliderThumbElement::registerForTouchEvents):
(WebCore::SliderThumbElement::unregisterForTouchEvents):
(WebCore::SliderThumbElement::hostDisabledStateChanged):
(WebCore::SliderThumbElement::shouldAcceptTouchEvents): Deleted.
* Source/WebCore/html/shadow/SliderThumbElement.h:
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::computeEventListenerRegionTypes):

Canonical link: <a href="https://commits.webkit.org/303558@main">https://commits.webkit.org/303558@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/667d8595d952e85059ac8d6b3caa9864377bd19a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132711 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5206 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43802 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140240 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/84738 "Built successfully") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/57f2e9be-f0d2-4115-aeb5-62955c355bd2) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/134581 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/5555 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4965 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101461 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/68758 "Passed tests") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/76fcd820-6135-482b-bee1-dbad1fc1edd5) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135657 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/3911 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118901 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82254 "Passed tests") | | [❌ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/af7195e4-6d83-4199-aaee-1cc10f0cde0f) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/3798 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/1466 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/83474 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/112770 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/37017 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142900 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4876 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37604 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109841 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4958 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4219 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110018 "Passed tests") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/27913 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3721 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115172 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/58358 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20589 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4930 "Built successfully") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/33519 "Found 1 new test failure: fast/mediastream/granted-denied-request-management1.html (failure)") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4768 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68381 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5021 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4887 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/html/HTMLInputElement.cpp:
```diff
@@ -636,7 +636,7 @@ inline void HTMLInputElement::runPostTypeUpdateTasks()
 
 #if ENABLE(TOUCH_EVENTS)
-inline void HTMLInputElement::updateTouchEventHandler()
+void HTMLInputElement::updateTouchEventHandler()
```
Is this perf neutral? (now that it no longer has the `inline` suggestion)

lilyspiniolas replied:
Not sure, I re-inlined it, though.

aprotyas commented on Source/WebCore/style/StyleAdjuster.cpp:
```diff
@@ -378,6 +378,13 @@ OptionSet<EventListenerRegionType> Adjuster::computeEventListenerRegionTypes(con
         findListeners(eventNames().gesturestartEvent, EventListenerRegionType::GestureStart, EventListenerRegionType::NonPassiveGestureStart);
     }
+
+    if (eventTarget.hasInternalTouchEventHandling()) {
```
Is this synonymous with synchronous event listener region types? Should we maybe consult the event target? (in case the `::Touch[...]` region types are ever appropriate)

lilyspiniolas replied:
Yes, it is synonymous with synchronous event listener region types. Previously, the switch and range controls used synchronous event listeners, and regions with unaccelerated overflow scrolling require synchronous touch event listeners as well.

lilyspiniolas committed:
[Site Isolation] Touch event regions are missing for elements which have touch handlers but no JS listeners
https://bugs.webkit.org/show_bug.cgi?id=302689
rdar://164939029

Reviewed by Abrar Rahman Protyasha.

Switch controls, slider thumb elements, and elements with unaccelerated overflow scrolling all
require synchronous touch event handling but do not have JS listeners. The lack of JS listeners
results in us failing to generate touch event regions for these elements.

Fix this issue by introducing a new EventTargetFlag `HasInternalTouchEventHandling` which
is set for these elements. In addition to targeting elements based on JS listeners in
`Adjuster::computeEventListenerRegionTypes`, we now look for the new EventTargetFlag as well.
If present, we add synchronous event listener region types NonPassiveTouchStart,
NonPassiveTouchMove, NonPassiveTouchEnd, & NonPassiveTouchCancel.

In `SliderThumbElement` we no longer check if a renderer is present before registering for
touch events. Elements without a renderer will not get an event region in the first place.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/EventTarget.h:
* Source/WebCore/html/CheckboxInputType.cpp:
(WebCore::CheckboxInputType::disabledStateChanged):
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::updateTouchEventHandler): Deleted.
* Source/WebCore/html/HTMLInputElement.h:
* Source/WebCore/html/HTMLInputElementInlines.h: Added.
(WebCore::HTMLInputElement::updateTouchEventHandler):
* Source/WebCore/html/InputType.cpp:
(WebCore::InputType::hasTouchEventHandler const):
* Source/WebCore/html/RangeInputType.cpp:
(WebCore::RangeInputType::createShadowSubtree):
* Source/WebCore/html/shadow/SliderThumbElement.cpp:
(WebCore::SliderThumbElement::willDetachRenderers):
(WebCore::SliderThumbElement::didAttachRenderers):
(WebCore::SliderThumbElement::registerForTouchEvents):
(WebCore::SliderThumbElement::unregisterForTouchEvents):
(WebCore::SliderThumbElement::hostDisabledStateChanged):
(WebCore::SliderThumbElement::shouldAcceptTouchEvents): Deleted.
* Source/WebCore/html/shadow/SliderThumbElement.h:
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::computeEventListenerRegionTypes):

Canonical link: https://commits.webkit.org/303558@main
